### PR TITLE
bpo-35900: Add a state_setter arg to save_reduce

### PR DESCRIPTION
**Pull request by pierreglaser**
Second PR of [bpo-35900](https://bugs.python.org/issue35900).

# What

The aim of this PR is to allow users to specify custom state setting methods for types they have no control on. This is done via adding a `state_setter` keyword argument in `Pickler.save_reduce`, expected to be a callable. This callable will be called at unpickling time with the following signature `state_setter(obj, state, slotstate)`.

# Why

A practical use case is the one of [cloudpickle](https://github.com/cloudpipe/cloudpickle) that provides extended serialization procedures, especially for functions and classes. In `cloudpickle`, functions and classes are reconstructed from scratch at unpickling time. As for other built-in types reconstructed from scratch (list, dicts...), circular-references are handled by initalizing a skeleton-object, and then updating it in a state-setting phase. This state-setting phase currently only relies on potential `__getstate__` methods, or fallbacks to a default procedure, and this is simply not enough to properly reconstruct functions and classes, hence this proposal.

See cloudpipe/cloudpickle#253 for more details.

- [x] implementaion 
- [x] unit-tests

# Notes

## dependencies

This PR comes alongside #12499 . Without it, leveraging #12499 changes won't work for `functions` and `classes`.

## cross-protocol/version usage
An interrogation that remains is whether using this new API with a protocol < 5 should raise an error at pickling time. If not, `unpickling` will not succeed if the pickle string is loaded using a `python < 3.8`.  This may be surprising for users as one can assume that the point of using low protocols is to ensure retro-compatibility with older python versions. On the other side, this API will probably be of interested for advanced user only, that would understand why using a new API with an old protocol. As of today, this is still an open question.

https://bugs.python.org/issue35900

### CODE REVIEW

File: Modules/_pickle.c [hunk 1]
```diff
@@ -3927,11 +3938,48 @@ save_reduce(PicklerObject *self, PyObject *args, PyObject *obj)
...
+
+            if (statetup == NULL) {
+                PyErr_SetString(st->PicklingError,
+                                "state must be either a dict or a tuple of"
+                                " length 2");
+                return -1;
+            }
+
+            if (save(self, statetup, 0) < 0 ||
+                _Pickler_Write(self, &build_op, 1) < 0){
```
Comment (ogrisel) [inline Modules/_pickle.c]: formatting: whitespace before `{`

File: Modules/_pickle.c [hunk 2]
```diff
@@ -6229,6 +6277,39 @@ load_build(UnpicklerObject *self)
         Py_INCREF(slotstate);
         Py_DECREF(tmp);
     }
+    /* proto 5 addition: state can embed a callable state setter */
```
Comment (ogrisel) [inline Modules/_pickle.c]: Is this really related to protocol 5? It does not seem to be the case: it can also be used with protocol 4 and lower.

File: Doc/library/pickle.rst [hunk 3]
```diff
@@ -629,6 +629,11 @@ or both.
      value``.  This is primarily used for dictionary subclasses, but may be used
      by other classes as long as they implement :meth:`__setitem__`.
 
+   * Optionally, an callable with a ``(obj, state)`` signature. This
```
Comment (ogrisel) [inline Doc/library/pickle.rst]: a callable

File: Doc/library/pickle.rst [hunk 4]
```diff
@@ -629,6 +629,11 @@ or both.
...
+     callable allows the user to control the state-updating part of classes
+     the user did not create. If not ``None``, this callable will have priority
```
Comment (ogrisel) [inline Doc/library/pickle.rst]: "the user did not create" does not seem very relevant.

May I suggest the following:

Optionally, a callable with a ``(obj, state)`` signature.  This makes it possible
to customize the way to object state is restored. If left to ``None``
(the default), the state will be updated with the object
:meth:`__setitem__` method instead.

File: Lib/test/pickletester.py [hunk 5]
```diff
@@ -3081,6 +3094,19 @@ def reduce_2(obj):
         self.assertEqual(default_load_dump(a), REDUCE_A)
         self.assertIsInstance(default_load_dump(b), BBB)
 
+        # End-to-end testing of save_reduce with the state_setter keyword
+        # argument. This is a dispatch_table test as state_setter is useful to
+        # register custom state-setting behavior for classes that were not
+        # created by the user.
```
Comment (ogrisel) [inline Lib/test/pickletester.py]: "for classes that were not created by the user."

=>

"when the default behavior of the `__setitem__` method is not flexible enough."

File: Modules/_pickle.c [hunk 6]
```diff
@@ -3927,11 +3938,48 @@ save_reduce(PicklerObject *self, PyObject *args, PyObject *obj)
...
-    }
+        if (state_setter == NULL) {
+            if (save(self, state, 0) < 0 ||
+                _Pickler_Write(self, &build_op, 1) < 0)
+                return -1;
+        }
+        else {
+            PyObject *statetup = NULL;
+
+
```
Comment (ogrisel) [inline Modules/_pickle.c]: style: remove at least one blank line.

File: Modules/_pickle.c [hunk 7]
```diff
@@ -3927,11 +3938,48 @@ save_reduce(PicklerObject *self, PyObject *args, PyObject *obj)
...
-        if (save(self, state, 0) < 0 ||
-            _Pickler_Write(self, &build_op, 1) < 0)
-            return -1;
-    }
+        if (state_setter == NULL) {
+            if (save(self, state, 0) < 0 ||
+                _Pickler_Write(self, &build_op, 1) < 0)
+                return -1;
+        }
+        else {
```
Comment (ogrisel) [inline Modules/_pickle.c]: style: the `else {` should be on the same line as the previous `}`

File: Modules/_pickle.c [hunk 8]
```diff
@@ -3927,11 +3938,48 @@ save_reduce(PicklerObject *self, PyObject *args, PyObject *obj)
...
+             * tempted to save a (state_setter, state) as state, but this would
+             * collide with load_build's (state, slotstate) special handling.
+             * Therefore, create a new format for state saving: (state_setter,
+             * state, slotstate)
+             */
+            if PyDict_Check(state)
+                statetup = Py_BuildValue("(OOO)", state_setter, state,
+                                         Py_None);
+            else if PyTuple_Check(state) {
+                if (PyTuple_GET_SIZE(state) == 2){
```
Comment (ogrisel) [inline Modules/_pickle.c]: style: missing whitespace before `{`.

File: Modules/_pickle.c [hunk 9]
```diff
@@ -6229,6 +6277,39 @@ load_build(UnpicklerObject *self)
...
+            statearg = Py_BuildValue("(OO)", state, slotstate);
+        }
+        else {
+            statearg = state;
+            Py_INCREF(state);
+        }
+
+        if (PyObject_CallFunctionObjArgs(setstate, inst, statearg,
+                                         NULL) == NULL){
+            return -1;
```
Comment (ogrisel) [inline Modules/_pickle.c]: You probably need to:

```
        Py_DECREF(setstate);
        Py_DECREF(state);
        Py_DECREF(statearg);
        Py_DECREF(slotstate);
```

here no?

Maybe `Py_DECREF(state)` and `Py_DECREF(slotstate)` should be moved up.

**Review by ogrisel — COMMENTED** [review]
Here is a batch of comments:

File: Doc/library/pickle.rst (hunk 4 above)
Comment (pierreglaser) [reply Doc/library/pickle.rst]: > meth:__setitem__ method instead.

s/`__setitem__`/`__setstate__`

> "the user did not create" does not seem very relevant.

Well, the usual way to restore an object state is to implement a `__setstate__` method. The only setting where one cannot use `__setstate__` is when `obj` is a foreign object, say from a class that was not defined by the user.

Actually, the `state_setter` argument is similar to the `dict_setitems` and the `list_setitems` argument, whose purpose is to modify the hard coded `state updating`  behavior (looping through all items and appending them) for `list` and `dict`.  Using `state_setter`, we modify the hard-coded behavior given in `__setstate__`, in our case for classes and functions, although `state_setter` will be called on any object, if provided.

File: Modules/_pickle.c (hunk 2 above)
Comment (pierreglaser) [reply Modules/_pickle.c]: For now it is not you are right. I do not really know whether we should enable this change only for protocol 5. Enabling this for lower protocol will generate cryptic errors when trying to read pickle string of recent pythons from older pythons (although doing this is not recommended anyways). 

Another example of a non-opcode-related protocol addition (not written by me)
https://github.com/python/cpython/blob/f5f336a819a3d881bb217bf8f9b5cacba03a4e45/Modules/_pickle.c#L6220-L6222

**Comment by pierreglaser** [comment]
ping @pitrou

File: Modules/_pickle.c [hunk 10]
```diff
@@ -6229,6 +6276,27 @@ load_build(UnpicklerObject *self)
...
         Py_DECREF(tmp);
     }
+    /* state can embed a callable state setter */
+    else if (PyTuple_Check(state) && PyTuple_GET_SIZE(state) == 3) {
+        PyObject *state_slotstate;
+
+        setstate = PyTuple_GET_ITEM(state, 0);
+        state_slotstate = PyTuple_GetSlice(state, 1, 3);
+
+        Py_INCREF(setstate);
```
Comment (ogrisel) [inline Modules/_pickle.c]: move this below the line with:

```c
setstate = PyTuple_GET_ITEM(state, 0);
```
?

File: Lib/pickle.py [hunk 11]
```diff
@@ -655,7 +655,24 @@ def save_reduce(self, func, args, state=None, listitems=None,
             self._batch_setitems(dictitems)
 
         if state is not None:
-            save(state)
+            msg = "state must be either a dict or a tuple of length 2"
+            if state_setter is None:
+                save(state)
+
+            elif isinstance(state, dict):
```
Comment (pitrou) [inline Lib/pickle.py]: I wonder if we wouldn't be better off with a separate opcode instead of reusing BUILD with a new argument syntax. This would put this change in the realm of pickle protocol 5 (making it harder for you to develop and test until PEP 574 is merged ;-)). To be discussed on python-dev? @ncoghlan

Comment (ogrisel) [reply Lib/pickle.py]: Actually @pierreglaser suggested IRL to use a nested reduce when we want to use a custom state_setter. This way we would not pervert the semantics of `BUILD` and the resulting pickle would be loadable in older Python versions.

Comment (pierreglaser) [reply Lib/pickle.py]: Yeah, I implemented this new method. It is fully backward compatible, but kind of hacky - I'll let you see by yourselves: https://github.com/pierreglaser/cpython/pull/2

File: Modules/_pickle.c (hunk 2 above)
Comment (ogrisel) [reply Modules/_pickle.c]: If pierreglaser#2 is merged (which I think it should), then this `state_setter` PR is really independent of the protocol version since a pickle generated using that new feature can be loaded by any older version of Python `pickle.load`.

File: Doc/library/pickle.rst [hunk 12]
```diff
@@ -629,6 +629,12 @@ or both.
      value``.  This is primarily used for dictionary subclasses, but may be used
      by other classes as long as they implement :meth:`__setitem__`.
 
+   * Optionally, a callable with a ``(obj, state)`` signature. This
+     callable allows the user to programatically control the state-updating
+     behavior of a specific object, instead of using ``obj``'s static
+     :meth:`__setstate__` method. If not ``None``, this callable will have
+     priority over ``obj``'s :meth:`__setstate__`.
+
```
Comment (ogrisel) [inline Doc/library/pickle.rst:637]: Please add a marker :

```
    .. versionadded:: 3.8
```

I am not sure about the required indentation level for the versionadded directive.

**Review by ogrisel — APPROVED** [review]
LGTM!

**Comment by pierreglaser** [comment]
OK @pitrou @ncoghlan how does this look to you now?

**Comment by ncoghlan** [comment]
I fixed a couple of small error message issues, and a typo in the news entry, and this otherwise looks OK to me.

However, I'll leave the final review to @pitrou :)

File: Doc/library/pickle.rst [hunk 13]
```diff
@@ -629,6 +629,15 @@ or both.
...
+     .. versionadded:: 3.8
+        The ``state_setter`` option was added.
```
Comment (pitrou) [inline Doc/library/pickle.rst]: It's not called `state_setter` (anymore?), so perhaps write `(obj, state)` instead?

Or, for example, more explicitly:
```rst
The optional sixth tuple item, ``(obj, state)``, was added.
```

Comment (pitrou) [reply Doc/library/pickle.rst]: Also, in line 601 above, the sentence """When a tuple is returned, it must be between two and five items long""" should be updated.

File: Lib/pickle.py [hunk 14]
```diff
@@ -655,8 +655,25 @@ def save_reduce(self, func, args, state=None, listitems=None,
...
+            if state_setter is None:
+                save(state)
+                write(BUILD)
+            else:
+                # If a state_setter is specified, call it instead of load_build
+                # to update obj's with its previous state.
+                # First, push state_setter and its tuple of expected arguments
+                # (obj, state) onto the stack.
+                save(state_setter)
+                save(obj)
```
Comment (pitrou) [inline Lib/pickle.py]: This will use the memo? Perhaps add a comment?

File: Lib/test/pickletester.py [hunk 15]
```diff
@@ -3081,6 +3094,20 @@ def reduce_2(obj):
         self.assertEqual(default_load_dump(a), REDUCE_A)
         self.assertIsInstance(default_load_dump(b), BBB)
 
+        # End-to-end testing of save_reduce with the state_setter keyword
+        # argument. This is a dispatch_table test as the primary goal of
+        # state_setter is to tweak objects reduction behavior.
+        # In particular, state_setter is useful when the default __setstate__
+        # behavior is not flexible enough.
+        def reduce_bbb(obj):
+            return BBB, (), obj.__dict__, None, None, setstate_bbb
```
Comment (pitrou) [inline Lib/test/pickletester.py:3114]: Can you make this a test with a class that also has a `__setstate__`, to validate that the state_setter has priority?

File: Misc/NEWS.d/next/Library/2019-03-27-15-09-00.bpo-35900.fh56UU.rst [hunk 16]
```diff
@@ -0,0 +1,2 @@
+Allow the user to specify custom functions to set the state of an object in
+save_reduce via the new state_setter keyword argument.
```
Comment (pitrou) [inline Misc/NEWS.d/next/Library/2019-03-27-15-09-00.bpo-35900.fh56UU.rst]: "save_reduce" and "state_setter" are implementation details, so perhaps phrase this differently?
For example:
```rst
Allow reduction methods to return a 6-item tuple where the 6th item specifies a custom state-setting method that's called instead of the regular ``__setstate__`` method.
```

File: Modules/_pickle.c [hunk 17]
```diff
@@ -3656,6 +3656,7 @@ save_reduce(PicklerObject *self, PyObject *args, PyObject *obj)
     PyObject *state = NULL;
     PyObject *listitems = Py_None;
     PyObject *dictitems = Py_None;
+    PyObject *state_setter = Py_None;
```
Comment (pitrou) [inline Modules/_pickle.c:3665]: Can write `PyObject* state_setter = NULL;` directly.

**Review by pitrou — CHANGES_REQUESTED** [review]
Looks mostly good to me. Just some nits.

File: Lib/pickle.py (hunk 14 above)
Comment (pierreglaser) [reply Lib/pickle.py]: I think this will always use the memo when saving obj from the public ``pickle``API, as IIRC in this case `obj` is always specified in `save_reduce`. However, I just figured that this code could actually fail is if a users directly calls the semi-private `save_reduce` while not specifying obj (``save(obj)`` will become ``save(None)``). Not sure this is a case we want to handle though. WDYT?

Comment (pitrou) [reply Lib/pickle.py]: No, calling `save_reduce` directly is not supported.

File: Modules/_pickle.c (hunk 17 above)
Comment (pierreglaser) [reply Modules/_pickle.c:3665]: Correct me if I am wrong, but I think we cannot, because the following check:
https://github.com/pierreglaser/cpython/blob/fabb59319f37848bdbb4c9a1951857cdd45ed75b/Modules/_pickle.c#L3713-L3715
will segfault.

Comment (pitrou) [reply Modules/_pickle.c:3665]: Fair enough. Also, it seems this allows an explicit `None` as 6th tuple item (which may be useful if some day a 7th item appears ;-)).

File: Modules/_pickle.c [hunk 18]
```diff
@@ -3708,6 +3710,15 @@ save_reduce(PicklerObject *self, PyObject *args, PyObject *obj)
         return -1;
     }
 
+    if (state_setter == Py_None)
+        state_setter = NULL;
+    else if (!PyFunction_Check(state_setter)) {
```
Comment (pitrou) [inline Modules/_pickle.c]: Ah, I overlooked this. `PyFunction_Check` only accepts pure Python functions. For arbitrary callables, you should use `PyCallable_Check` instead...

**Review by pitrou — CHANGES_REQUESTED** [review]
Sorry, one issue I overlooked.

File: Modules/_pickle.c (hunk 18 above)
Comment (pierreglaser) [reply Modules/_pickle.c]: Ah, did not know about that. Fixed.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
I can reproduce the cert failures on git master, so they're unrelated.

**Comment by pitrou** [comment]
The test failure was fixed on master, can you merge / rebase?

**Comment by pierreglaser** [comment]
Thanks a lot @pitrou!
